File: src/543DiameterOfBinaryTree/543DiameterOfBinaryTree.cpp
#include <iostream>
using namespace std;

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
  int maxDepth(TreeNode *root) {
    if (root == nullptr)
      return 0;
    return 1 + max(maxDepth(root->left), maxDepth(root->right));
  }

public:
  int diameterOfBinaryTree(TreeNode *root) {
    if (root == nullptr)
      return 0;
    return max(max(maxDepth(root->left) + maxDepth(root->right),
                   diameterOfBinaryTree(root->left)),
               diameterOfBinaryTree(root->right));
  }
};

int main() { return 0; }

File: src/543DiameterOfBinaryTree/543DiameterOfBinaryTree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn diameter_of_binary_tree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        Self::diameter_of_binary_tree_helper(&root)
    }
    
    pub fn diameter_of_binary_tree_helper(root: &Option<Rc<RefCell<TreeNode>>>) -> i32 {
        if let Some(tn) = root {
            let mut node = tn.borrow();
            let left_depth = Self::max_depth(&node.left);
            let right_depth = Self::max_depth(&node.right);
            let depth = left_depth + right_depth;
            std::cmp::max(std::cmp::max(depth, Self::diameter_of_binary_tree_helper(&node.left)), Self::diameter_of_binary_tree_helper(&node.right))
        }else{
            0
        }
    }
    
    pub fn max_depth(root: &Option<Rc<RefCell<TreeNode>>>) -> i32 {
        if let Some(tn) = root {
            let mut node = tn.borrow();
            1 + std::cmp::max(Self::max_depth(&node.left), Self::max_depth(&node.right))  
        }else{
            0
        }
    }
}
